libcalico-go/lib/epstatusfile/status_file_watcher.go

```diff
@@ -0,0 +1,245 @@
+		lastState:              make(map[string]os.FileInfo),
+		stopChan:               make(chan struct{}),
+		pollTicker:             pollTicker,
+		pollC:                  make(chan struct{}, 1),
+		newFsnotifyWatcherFunc: newFsnotifyWatcherFunc,
+	}
+}
+
+func (w *FileWatcher) SetCallbacks(callbacks Callbacks) {
+	w.callbacks = callbacks
+}
+
+func (w *FileWatcher) Start() {
+	go w.runWatcher()
+
+}
+
+func (w *FileWatcher) newFsnotifyWatcher() error {
+	// reset w.fsWatcher
+	w.fsWatcher = nil
+
+	watcher, err := w.newFsnotifyWatcherFunc()
+	if err != nil {
+		log.WithError(err).Error("Error initializing fsnotify.")
+		return err
+	}
+
+	err = watcher.Add(w.dir)
+	if err != nil {
+		log.WithError(err).Error("Error adding directory to fsnotify.")
+		return err
+	}
+
+	w.fsWatcher = watcher
+	log.WithField("dir", w.dir).Info("Started watching directory via fsnotify.")
+	return nil
+}
+
+func (w *FileWatcher) runFsnotifyWatcher(watcher *fsnotify.Watcher) error {
+	w.fsnotifyActive = true
+
+	// Get current state of the directory and emit initial events.
+	w.scanDirectory()
+
+	// Listen for events and loop until error occurs.
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				// Stop if channel is closed.
+				return fmt.Errorf("fsnotify channel is closed unexpectly")
+			}
+			log.WithFields(log.Fields{
+				"file": event.Name,
+				"op":   event.Op,
+			}).Debug("received file events")
+
+			filePath := event.Name
+			if event.Op == fsnotify.Remove {
+				w.callbacks.OnFileDeletion(filePath)
+				delete(w.lastState, filePath)
+			} else {
+				fileInfo, err := os.Stat(filePath)
+				if err != nil {
+					log.WithError(err).Error("Failed to get file info on a fsnotify event.")
+				} else if !fileInfo.IsDir() {
+					if event.Op == fsnotify.Create {
+						w.lastState[filePath] = fileInfo
+						w.callbacks.OnFileCreation(filePath)
+					}
+					if event.Op == fsnotify.Write {
+						w.lastState[filePath] = fileInfo
+						w.callbacks.OnFileUpdate(filePath)
+					}
+				}
+			}
+
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				// Stop if channel is closed.
+				return fmt.Errorf("fsnotify channel is closed unexpectly")
+			}
+			log.WithError(err).Error("fsnotify error. Falling back to polling.")
+			return err
+		case <-w.stopChan:
+			return nil
+		}
+	}
+}
+
+// Start begins watching the directory.
+func (w *FileWatcher) runWatcher() {
+	defer func() {
+		if w.fsWatcher != nil {
+			defer w.fsWatcher.Close()
+		}
+	}()
+
+	for {
+		// Try to get a fsnotify watcher if possible.
+		err := w.newFsnotifyWatcher()
+		if err != nil {
+			log.WithError(err).Info("Error initializing fsnotify. Falling back to polling.")
+		}
+
+		if w.fsWatcher != nil {
+			// Run fsnotify watcher loop if possible.
+			err := w.runFsnotifyWatcher(w.fsWatcher)
+			w.fsnotifyActive = false
+			if err != nil {
+				// fall back to polling.
+				log.WithError(err).Info("Start polling directory on updates.")
+				w.pollC <- struct{}{}
```

Comment:
I think this send could deadlock:
* First time through the loop, send is OK, leaves pending item on the channel.
* Then `w.pollTicker.C` pops in the select below and select chooses that instead of pollC.
* Loop and hit this error again, now we deadlock.
Not quite sure what this is trying to do.  Could you just `continue` to tool again?  Or should you let it fall through and wait for the ticker?

Reply:
That's a mistake. I should just let it fall through and wait for the ticker. There's no need for this channel.